File: Leetcode/C++/108. Convert Sorted Array to Binary Search Tree/solution.cpp
// https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/description/

#include <vector>
#include <iostream>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *sortedArrayToBST(vector<int> &nums)
    {
        return makeBST(nums, 0, nums.size());
    }

    TreeNode *makeBST(vector<int> &nums, int left, int right)
    {
        if (left >= right)
        {
            return NULL;
        }
        // Prevent overflow
        int mid = left + (right - left) / 2;
        TreeNode *root = new TreeNode(nums[mid]);
        // cout << left << " " << right << " " << nums[mid] << endl;    Debuging code
        root->left = makeBST(nums, left, mid);
        root->right = makeBST(nums, mid + 1, right);

        return root;
    }
};

File: Leetcode/Java/108. Convert Sorted Array to Binary Search Tree/solution.java
// https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/description/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return convert(nums, 0, nums.length - 1);
    }

    private TreeNode convert(int[] nums, int start, int end) {
        if (start > end) {
            return null;
        }
        // To create a height-balanced tree, the left and right sides should have as
        // similar a number of nodes as possible. Therefore, set the middle of the array
        // as the root and divide the left and right sides.
        int mid = start + (end - start) / 2;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = convert(nums, start, mid - 1);
        root.right = convert(nums, mid + 1, end);

        return root;
    }
}